### PR TITLE
Allow ref counted keys and values

### LINKED ISSUE #809: Allow using `Arc<str>` in key-value strings
Currently, `opentelemetry` represents strings in key-value pairs (i.e. `Key`s and `Value::String`s) using `Cow<'static, str>`. This has the advantage that when a string constant is used, it isn't necessary to allocate a `String` on the heap for it. If a string is dynamically constructed at runtime, a `String` is used instead.

One deficiency of this approach is that oftentimes, a string may be dynamically determined at runtime, but repeated many times in an application's traces. In this case, such a string is represented as a `String`, and cloning it requires allocating memory and copying all of the string's bytes into that allocation. This has an unfortunate performance cost, especially when the same string is repeated a large number of times. It would be nice if it was also possible to use an `Arc<str>` to represent strings in key-value pairs. That way, cloning a repeated string to emit it as part of a trace would only require incrementing a reference count, reducing the overhead of using the string (and decreasing overall memory use).

This came up while reviewing a change to the `tracing` crate's OpenTelemetry integration (https://github.com/tokio-rs/tracing/pull/2135#discussion_r878597135).

I think it should be possible to support `Arc<str>` as a representation for `Key`s without making a breaking change, since `Key` is a newtype whose field is private. Although the `Key::new` constructor takes `S: Into<Cow<'static, str>>`, we could add a new `from_arc` (or similar) constructor. Then, the internal representation of `Key` could be changed to something like this:
```rust
pub struct Key(KeyInner);

enum KeyInner {
    Static(&'static str),
    Owned(String),
    RefCounted(Arc<str>),
}

impl Key {
    pub fn new<S: Into<Cow<'static, str>>>(value: S) -> Self {
        Self(match value.into() {
             Cow::Owned(s) => KeyInner::Owned(s),
             Cow::Borrowed(s) => KeyInner::Static(s),
        })
    }

    pub fn from_arc(value: Arc<str>) -> Self {
         Self(KeyInner::RefCounted(value))
    }
    
    // ...
}
```

Unfortunately, though, a similar change for `Value` isn't possible without breaking API compatibility, because `Value` is a `pub enum` without `#[non_exhaustive]`.

### PR DESCRIPTION
**Pull request by jtescher**
This updates attribute `Key` and `Value` inner types from `Cow<'static, str>` to a new private `OtelString` enum to support static, owned, or ref counted values.

Resolves #809

### CODE REVIEW

File: opentelemetry-contrib/Cargo.toml [hunk 1]
```diff
@@ -34,7 +34,7 @@ serde_json = { version = "1", optional = true}
 futures = {version = "0.3", optional = true}
 async-trait = {version = "0.1", optional = true}
 tokio = { version = "1.0", features = ["fs", "io-util"], optional = true }
-async-std = { version = "1.6", optional = true }
+async-std = { version = "1.10", optional = true }
```
Comment (TommyCpp) [inline opentelemetry-contrib/Cargo.toml:37]: Although it probably won't break the MSRV, we should probably pin this to 1.10 to be consistent with other crates?

**Review by TommyCpp — APPROVED** [review]
I think to pin the `indexmap` to `=1.8` should fix the MSRV. I will spend some time and see if we can remove the `indexmap` as a dependency. Some context of this change can be found in https://github.com/open-telemetry/opentelemetry-rust/pull/799

File: opentelemetry-api/src/common.rs [hunk 2]
```diff
@@ -166,11 +272,75 @@ pub enum Value {
...
+            OtelString::Owned(s) => s.fmt(f),
+            OtelString::Static(s) => s.fmt(f),
+            OtelString::RefCounted(s) => s.fmt(f),
+        }
+    }
+}
+
+impl StringValue {
+    /// Returns a string slice to this value
+    pub fn as_str(&self) -> &str {
```
Comment (djc) [inline opentelemetry-api/src/common.rs:302]: Should probably also implement `AsRef<str>`.

File: opentelemetry-sdk/src/resource/process.rs [hunk 3]
```diff
@@ -25,13 +24,10 @@ impl ResourceDetector for ProcessResourceDetector {
...
-            .map(|arg| Cow::from(arg.to_string_lossy().into_owned()))
-            .collect::<Vec<Cow<'_, str>>>();
+            .map(|arg| arg.to_string_lossy().into_owned())
+            .collect::<Vec<String>>();
         Resource::new(vec![
-            KeyValue::new(
-                "process.command_args",
-                Value::Array(Array::String(cmd_arg_val)),
-            ),
+            KeyValue::new("process.command_args", Value::Array(cmd_arg_val.into())),
```
Comment (djc) [inline opentelemetry-sdk/src/resource/process.rs:30]: For example here we first allocate a `Vec<String>` and then `cmd_arg_val.into()` will allocate a new `Vec<StringValue>`, redoing the outer allocation.

File: opentelemetry-api/src/common.rs [hunk 4]
```diff
@@ -153,9 +226,42 @@ into_array!(
     (Vec<bool>, Array::Bool),
     (Vec<i64>, Array::I64),
     (Vec<f64>, Array::F64),
-    (Vec<Cow<'static, str>>, Array::String),
+    (Vec<StringValue>, Array::String),
 );
 
+impl From<Vec<&'static str>> for Array {
+    /// Convenience method for creating a `Value` from a `&'static str`.
```
[the remark this answers is not in the document]
Comment (jtescher) [reply opentelemetry-api/src/common.rs:233]: Happy to remove them

File: opentelemetry-sdk/src/resource/process.rs (hunk 3 above)
Comment (jtescher) [reply opentelemetry-sdk/src/resource/process.rs:30]: Agreed

File: opentelemetry-contrib/Cargo.toml (hunk 1 above)
Comment (jtescher) [reply opentelemetry-contrib/Cargo.toml:37]: This was only to get CI passing. Still have to resolve the MSRV issues on main.

**Review by djc — APPROVED** [review]

**Pull request merged**